traversal/expression/ExpressionFactory.java

```diff
@@ -0,0 +1,110 @@
+    private static final Map<com.vaticle.typedb.core.encoding.Expression.Operation, OperationFactory> OPERATIONS = Operation.createRegistry();
+    private static final Map<com.vaticle.typedb.core.encoding.Expression.Function, FunctionFactory> FUNCTIONS = Function.createRegistry();
+    private static final Expression<?>[] toArrayTemplate = new Expression[0];
+
+    public interface OperationFactory<T1 extends Expression<?>, T2 extends Expression<?>> {
+        Expression<?> build(T1 x1, T2 x2);
+    }
+
+    public interface FunctionFactory {
+        Expression<?> build(Expression<?>[] args);
+    }
+
+    public static Expression<?> build(ValueConstraint.Assignment.Expression expr,
+                                      Map<ValueConstraint.Assignment.Expression, com.vaticle.typedb.core.encoding.Expression> resolvedExpressions) {
+        if (expr.isConstant()) {
+            if (expr.asConstant().isBoolean()) return instantiateConstant(expr.asConstant().asBoolean().value());
+            else if (expr.asConstant().isLong()) return instantiateConstant(expr.asConstant().asLong().value());
+            else if (expr.asConstant().isDouble()) return instantiateConstant(expr.asConstant().asDouble().value());
+            else if (expr.asConstant().isString()) return instantiateConstant(expr.asConstant().asString().value());
+            else if (expr.asConstant().isDateTime()) return instantiateConstant(expr.asConstant().asDateTime().value());
+            else throw TypeDBException.of(ILLEGAL_STATE);
+        } else if (expr.isThingVar()) {
+            assert expr.asThingVar().variable().reference().isName();
+            return instantiateThingVar(resolvedExpressions.get(expr).type(), Identifier.Variable.of(expr.asThingVar().variable().reference().asName()));
+        } else if (expr.isValueVar()) {
+            return instantiateValueVar(resolvedExpressions.get(expr).type(), Identifier.Variable.of(expr.asValueVar().variable().reference().asValue()));
+        } else if (expr.isBracketed()) {
+            return build(expr.asBracketed().nestedExpression(), resolvedExpressions);
+        } else if (expr.isOperation()) {
+            Pair<ValueConstraint.Assignment.Expression, ValueConstraint.Assignment.Expression> operands = expr.asOperation().operands();
+            return instantiateOperation(resolvedExpressions.get(expr), build(operands.first(), resolvedExpressions), build(operands.second(), resolvedExpressions));
+        } else if (expr.isFunction()) {
+            List<Expression<?>> args = new ArrayList<>();
+            iterate(expr.asFunction().arguments()).map(argExpr -> build(argExpr, resolvedExpressions)).forEachRemaining(args::add);
+            return instantiateFunction(resolvedExpressions.get(expr), args);
+        } else throw TypeDBException.of(ILLEGAL_STATE);
+    }
+
+    public static Constant.Boolean instantiateConstant(java.lang.Boolean value) { return new Constant.Boolean(value); }
```

Comment:
you don't need `instantiate` on a static method, these are factory methods already

Reply:
So just `constant`?

Reply:
Yep!